### PR TITLE
Elements - Forms

### PR DESCRIPTION
**Pull request by clrux**
This work adds forms and form elements to the pattern library. It consists of:

* input, select, and textarea styles
* scripts that visually replace select, radio, and checkbox controls giving us more control over the UI
* suggested, semantic markup structure
* styling for form elements in general

### Form control replacements

For certain `inputs` and `selects` we're offering options to use either the default browser-rendered control or a replaced control, either rendered by JS or CSS or both. The idea for the radio buttons and checkboxes came from [Heydon Pickering and an article on Sitepoint](http://www.sitepoint.com/replacing-radio-buttons-without-replacing-radio-buttons/). The form controls with options are:
* `radio` buttons
* `checkbox` buttons
* `select` menus

#### Radio buttons and checkboxes
To enable the enhanced version of a radio button or a checkbox, do the following:

 `<input type="radio" class="replace-radio">`
`<input type="checkbox" class="replace-checkbox">`

Replacing radio buttons and checkboxes uses an _accessible_ CSS-only approach that hides the default control, replacing it with a CSS rendered one. Interactions with the control still happen on the default control which maintains full accessibility and browser methods.

#### Select boxes
To enable the enhanced version of a select menu, do the following:

`<select class="replace-select">`

Replacing select menus uses an _accessible_ JavaScript and CSS approach. JavaScript hides the default control, replacing it with a styled wrapper using the selected value. Interaction still happens with the default control which maintains complete accessibility and compliance.

### States

For each of the form input controls we've created and demonstrated the following states:
* `normal`
* `active`, `checked`, `selected`, and `hovered`
* `disabled`
* `required`

### User feedback, message handling

This piece of work also includes user feedback messages for form field validation. The cases included are:
* `warning`
* `error`
* `success`

### Nested styles

In many cases we're using nested styles. The aim here is to slightly reduce modularity but promote a semantic structure.

### Form control heights and widths

To provide a nice set of options both for us internally and the Open edX community, I've created various sizes for form controls.

#### Height
Height primarily relates to `textareas`. To use a height, choose one of the following:
* `short`
* `base` our default height
* `tall`

#### Width
Width can be set for any form control. To use a width, choose one of the following:
* `short`
* `base` our default width
* `wide`
* `block` which fills the entire width of the parent container

### Range slider and Progress bars

The visual UI included a slider, but it was a hybrid of a progress bar and a range slider. Since nothing like that is available natively I created two different elements: `input type="range"` and `progress` and created styles for these as I found fit.

#### Slider/Range

The slider control allows "notched" or fixed selections along a track. There is no "before/after" or "complete/incomplete" which was the discrepancy in the UI. I've included styles for this that makes the handle increase in size when it's hovered over, making the click target/control more usable.

#### Progress

Unlike slider controls, progress bars _do_ have a "before/after" or a "complete/incomplete" state. This matches the UI a little closer, but progress bars don't have drag handles.

---

## Browser testing
This work has been tested in the following browsers:

| Platform | Browser | Version | Passes |
|----------|---------|---------|--------|
| Mac      | Chrome  | 44      | Yes    |
|          | Firefox | 24.3    | Yes    |
|          | Safari  | 8       | Yes    |
| PC       | Chrome  | 43      | Yes    |
|          | Firefox | 37      | Yes    |
|          | IE      | 10      | Yes *  |
|          |         | 11      | Yes *  |

* Mostly works. See [this article](http://brennaobrien.com/blog/2014/05/style-input-type-range-in-every-browser.html) for more information.

---
__Reviewers:__
- [x] * @talbs
- [x] * @frrrances 
- [x] * @AlasdairSwan  
- [x] * @cptvitamin

---

_This PR replaces #75. For history, comment, etc. please see it for details._

### CODE REVIEW

File: _src/pattern-library/sass/components/_forms.scss [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+    &::-ms-fill-lower {
+        @include input-track();
+    }
+
+    // IE: notches along track
+    &::-ms-track {
+        border-width: rem($baseline/2); // the handle is clipped without this
+        border-color: transparent; // the handle is clipped without this
+        background: transparent; // the handle is clipped without this
+        color: transparent; // hides the notches along the track; remove if you want them
```
Comment (frrrances) [inline _src/pattern-library/sass/components/_forms.scss]: Nice notes.

File: _posts/elements/2015-05-12-forms.md [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+                    <option value="Orange">Orange</option>
+                    <option value="Yellow">Yellow</option>
+                    <option value="Green">Green</option>
+                    <option value="Blue" selected="selected">Blue</option>
+                    <option value="Indigo">Indigo</option>
+                    <option value="Violet">Violet</option>
+                </select>
+            </div>
+            <div class="field">
+                <label class="field-label" for="select_02_suc">Choose your favorite color of the rainbow (enhanced select and disabled menu)</label>
```
Comment (frrrances) [inline _posts/elements/2015-05-12-forms.md]: The for value is the same as the above select "select_02_suc" - not sure if that matters in this example?

File: _posts/elements/2015-05-12-forms.md [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+                <select class="field-input input-select replace-select" id="select_02_suc" name="select_02_suc" disabled>
```
Comment (frrrances) [inline _posts/elements/2015-05-12-forms.md]: The ID and name value is the same as the above select "select_02_suc" - not sure if that matters in this example?

**Comment by frrrances** [comment]
Thanks for the good work @clrux! One small question, but not a blocker. :+1:

**Comment by clrux** [comment]
@frrrances Good catch on that duplicate `id`/`for`, thanks. I've changed them to match the field appropriately and be unique.

**Pull request merged**